GNX/man/man3/walkdir.3
.ig
	@(#)walkdir.3	1.3	11/3/83
	@(#)Copyright (C) 1983 by National Semiconductor Corp.
..
.TH WALKDIR 3
.SH NAME
walkdir \- recursively find files in a directory hierarchy
.SH SYNOPSIS
.nf
.B #include <sys/types.h>
.B #include <dir.h>
.B #include <stat.h>
.B #include <walkdir.h>
.PP
.B walkdir(startpath, calledroutine, flags)
.B char *startpath;
.B int (*calledroutine)();
.PP
.B calledroutine(path, statbuf, error)
.B char *path;
.B struct stat *statbuf;
.fi
.SH DESCRIPTION
.I Walkdir
provides a way to examine all files downward from a given file specification,
by recursively examining directories and returning all file names found.
As
.I walkdir
finds each file name,
it invokes
.IR calledroutine ,
supplying the file name as the first argument.
.PP
.I Startpath
is the path where the search starts.
If the specified path names a file other than a directory,
.I walkdir
will return only that file's name.
If the path is a directory,
then the names of all files in that directory will be returned,
including those in all lower directories.
For example,
to process the name of every file in the file system,
.I startpath
would be "/".
.PP
.I Calledroutine
is the user-defined routine which will be called for each file found.
It has three arguments.
.I Path
is the path to the current file, and can be used to reference the found file.
.I Statbuf
is a pointer to the status of the file (see
.IR stat (2)).
.I Error
is a flag indicating if the given file specification cannot be referenced
(one common error is a directory that is unreadable).
.I Calledroutine
should return zero to proceed to the next file.  If it returns nonzero,
.I walkdir
will abort the examination of the current directory.
This can be used to
skip over uninteresting directories.
.I Path
and
.I statbuf
are only valid during the call to
.IR calledroutine ,
and are overwritten between calls, so that they must be copied if they
are needed later.
.PP
.I Flags
specify the exact way in which the search is to take place, and consists of
one or more of the following bits ORed together
(defined in
.IR /usr/include/walkdir.h ).
.PP
.nf
#define SD_RETDIRF	0x01	/* return directories before files */
#define SD_RETDIRL	0x02	/* return directories after files */
#define SD_RETFILE	0x04	/* return files */
#define SD_RETERR	0x08	/* return files with errors */
#define SD_TYPERR	0x10	/* type error messages */
.fi
.PP
To return directories, one of SD_RETDIRF or SD_RETDIRL must be set.
To return non-directories, SD_RETFILE must be set.
To have inaccessible
files returned, SD_RETERR must be set (if not set,
.I error
does not need examination).
If SD_TYPERR is set, then
.I walkdir
will automatically type error messages to standard error
for inaccessible files.
.SH "SEE ALSO"
directory(3S),
scandir(3)
.SH DIAGNOSTICS
.I Walkdir
returns zero if all files were successfully examined.
Otherwise,
it returns the number of inaccessible files,
or -1 if
.I startpath
itself was inaccessible.


GNX/src/lib/libndir/walkdir.c
/*
 * walkdir.c: version 1.6 of 8/23/83
 * 
 */
# ifdef SCCS
static char *sccsid = "@(#)walkdir.c	1.6 (NSC) 8/23/83";
# endif

/* The walkdir routine provides a way of recursively examining all files
 * downward from a starting path specification.  For each file that is found,
 * a procedure supplied by the caller is called with the following arguments:
 *
 *	routine(path, statbuf, error)
 *
 * where path is the pathname to the current file, statbuf is the address of
 * a stat buffer for the file (or zero if the data was not available), and
 * error is nonzero if the current file could not be accessed.  If the path
 * name and status buffer are to be preserved, they should be copied, since
 * the path name lives on the stack, and the stat buffer is reused.  The
 * routine returns zero to continue looking at files, or nonzero to indicate
 * that the current directory is to be aborted.
 *
 * Walkdir accepts the starting path specification, the routine to be called
 * as described above, and flags describing how to carry out the search.
 * Walkdir returns zero if all files were accessable, -1 if the initial path
 * specification was not accessable, or a positive number giving the number
 * of files which were not accessable.
 */

#include <sys/types.h>
#include <sys/stat.h>
#include <dir.h>
#include <walkdir.h>

#define	READ	0		/* read function */
#define	MAXNAME	1000		/* maximum length of directory spec */
#define	MAXREAD	128		/* number of directory entries to read */

static	struct	stat	statbuf;	/* status buffer */
static	int	(*routine)();		/* routine to call */
static	int	flags;			/* flags */
static	int	errcount;		/* error count */
static	char	*path;			/* pointer to current path */
extern	int	alphasort();		/* sort routine for scandir */


walkdir(rootpath, userproc, opflags)
	char	*rootpath;		/* starting path */
	int	(*userproc)();		/* procedure to call */
	int	opflags;		/* flags (defined in walkdir.h) */
{
	char	pathdata[MAXNAME];	/* path specification */

	strcpy(pathdata, rootpath);
	routine = userproc;
	path = pathdata;
	flags = opflags;
	errcount = 0;
	if (stat(path, &statbuf)) {
		doerror(0);
		return(-1);
	}
	if ((statbuf.st_mode & S_IFMT) == S_IFDIR)
		dodir(path + strlen(path));
	else
		if (flags & SD_RETFILE) (*routine)(path, &statbuf, 0);
	return(errcount);
}


/* Recursively read a directory and handle its files */
static dodir(endpath)
	char	*endpath;		/* end of current path */
{
	register struct	direct	**dp;	/* directory array pointer */
	register struct	direct	**enddp;/* end of directory data */
	register struct	direct	*de;	/* current directory entry */
	register char	*cp;		/* pointer to char */
	struct	direct	**namelist;	/* list of files returned by scandir */
	int	ndir;			/* number of files in directory */
	struct	stat	dirstat;	/* saved status of directory */

	if (flags & SD_RETDIRF) {
		if ((*routine)(path, &statbuf, 0)) return;
	}
	ndir = scandir(path, &namelist, 0, alphasort);
	if (ndir < 0) {
		doerror(0);
		return;
	}
	dirstat = statbuf;
	enddp = namelist + ndir;
	for (dp = namelist; dp < enddp; dp++) {
		de = *dp;
		if ((de->d_ino == 0) ||
			((de->d_name[0] == '.') &&
			((de->d_name[1] == '\0') ||
				((de->d_name[1] == '.') &&
				(de->d_name[2] == '\0'))))) continue;
		cp = endpath;
		if (cp[-1] != '/') *cp++ = '/';
		strcpy(cp, de->d_name);
		if (stat(path, &statbuf)) {
			doerror(0);
			continue;
		}
		if ((statbuf.st_mode & S_IFMT) == S_IFDIR) {
			dodir(cp + strlen(cp));
			continue;
		}
		if (flags & SD_RETFILE) {
			if ((*routine)(path, &statbuf, 0)) break;
		}
	}
	for (dp = namelist; dp < enddp; dp++) free(*dp);
	free(*namelist);
	*endpath = '\0';
	if (flags & SD_RETDIRL) (*routine)(path, &dirstat, 0);
}


/* Here on an error */
static doerror(statbuf)
	struct	stat	*statbuf;	/* stat for file in error */
{
	if (flags & SD_TYPERR) perror(path);
	if (flags & SD_RETERR) (*routine)(path, statbuf, 1);
	errcount++;
}
